HW1/P6/P6.py

```diff
@@ -0,0 +1,91 @@
+from pyspark import SparkContext
+from collections import Counter
+import numpy as np
+
+#initialize spark
+sc = SparkContext('local', "Shakespeare")
+
+#get the wordlist
+wordlist = "/home/zhiqian/Dropbox/CS205/Homework/HW2_data/pg100.txt"
+
+
+
+word_data = sc.textFile(wordlist)
+
+word_data_split = word_data.map(lambda x: x.split(" "))
+word_data_split2 = word_data_split.flatMap(lambda x: x)
+print word_data_split2.take(10)
+print word_data_split2.count()
+
+#filter out the three conditions stated in the question
+word_data_filtered = word_data_split2.filter(lambda x: (not str(x).isdigit()) 
+											& (not all(map(str.isupper, str(x))))
+											& (not (str(x).endswith('.') & all(map(str.isupper, str(x)[:-1]) )    )))
+
+#give each word an index
+word_data_first = word_data_filtered.zipWithIndex()
+word_data_second = word_data_filtered.zipWithIndex()
+word_data_third = word_data_filtered.zipWithIndex()
+
+#change the index to the key, and stagger the index for second and third word
+word_data_first_1 = word_data_first.map(lambda (x,y): (y,str(x)))
+word_data_second_1 = word_data_second.map(lambda (x,y) : (y-1, str(x)))
+word_data_third_1 = word_data_third.map(lambda (x,y) : (y-2, str(x)))
+
+#filter out negative values
+word_data_second_2 = word_data_second_1.filter(lambda (x,y): x >= 0)
+word_data_third_2 = word_data_third_1.filter(lambda (x,y): x>= 0)
+
+#join words based on index
+first_two_words = word_data_first_1.join(word_data_second_2)
+first_three_words = first_two_words.join(word_data_third_2)
+
+#get rid of the index
+first_three_words_2 = first_three_words.map(lambda (x, ((y,z),w)): ((y,z), w))
+
+#using the first two words as key, groupbyKey
+first_three_words_3 = first_three_words_2.groupByKey().mapValues(lambda x: list(x))
+
+#count the elements in the list (third word)
+first_three_words_4 = first_three_words_3.map(lambda ((x,y), z): ((x,y), Counter(z)))
+
+key = (('Now','is'))
+result = first_three_words_4.map(lambda x: x).lookup(key)
+#result = first_three_words_4.filter(lambda ((x,y), z): (x == 'Now') & (y == 'is'))
+print result
+
+
+#----Now, to generate the Markov chain!-------------------------------
+
+for counter in range(0,10):
+	final_word_list = []
+	#take a random sample. NOTE: i am working with the list of 3rd words BEFORE they are counted
+	#as it is easier to take a random choice from there
```

Comment:
I believe the random choice should be weighted based on the counts of the words, since intuitively, that's how a Shakespearean sentence should be constructed (a word that Shakespeare uses 100 times shouldn't have the same probability of getting chosen as a word used 2 times). A PDF of the word counts should suffice.